espnet/nets/pytorch_backend/transducer/blocks.py

```diff
@@ -0,0 +1,280 @@
+from espnet.nets.pytorch_backend.transformer.subsampling import Conv2dSubsampling
+
+
+def build_input_layer(
+    input_layer,
+    idim,
+    odim,
+    pos_enc_class,
+    dropout_rate_embed,
+    dropout_rate,
+    pos_dropout_rate,
+    padding_idx,
+):
+    """Build input layer.
+
+    Args:
+        input_layer (str): input layer type
+        idim (int): input dimension
+        odim (int): output dimension
+        pos_enc_class (class): PositionalEncoding or ScaledPositionalEncoding
+        dropout_rate_embed (float): dropout rate for embedding
+        dropout_rate (float): dropout rate
+        pos_dropout_rate (float): dropout rate for positional encoding
+        padding_idx (int): padding index for embedding
+
+    Returns:
+        (*): input layer
+
+    """
+    if input_layer == "linear":
+        return torch.nn.Sequential(
+            torch.nn.Linear(idim, odim),
+            torch.nn.LayerNorm(odim),
+            torch.nn.Dropout(dropout_rate),
+            torch.nn.ReLU(),
+            pos_enc_class(odim, pos_dropout_rate),
+        )
+    elif input_layer == "conv2d":
+        return Conv2dSubsampling(idim, odim, dropout_rate)
+    elif input_layer == "vgg2l":
+        return VGG2L(idim, odim)
+    elif input_layer == "embed":
+        return torch.nn.Sequential(
+            torch.nn.Embedding(idim, odim, padding_idx=padding_idx),
+            pos_enc_class(odim, pos_dropout_rate),
+        )
+    elif input_layer == "t-linear":
+        return torch.nn.Linear(idim, odim)
+    elif input_layer == "c-embed":
+        return torch.nn.Sequential(
+            torch.nn.Embedding(idim, odim, padding_idx=padding_idx),
+            torch.nn.Dropout(dropout_rate_embed),
+        )
+    elif input_layer is None:
+        return pos_enc_class(odim, pos_dropout_rate)
+    else:
+        raise NotImplementedError("Support: linear, conv2d, vgg2l and embed")
+
+
+def build_transformer_layer(
+    transformer_layer_class,
+    block_arch,
+    pw_layer_type,
+    pw_conv_kernel_size,
+    pos_dropout_rate,
+    att_dropout_rate,
+    dropout_rate,
+):
+    """Build function for Transformer layer.
+
+    Args:
+        transformer_layer_class (class): whether EncoderLayer or DecoderLayer
+        block_arch (dict): layer main arguments
+        pos_layer_type (str): positionwise layer type
+        pos_conv_kernel_size (int) : kernel size for positionwise conv1d layer
+        pos_dropout_rate (float): dropout rate for positional encoding
+        att_dropout_rate (float): dropout rate for attention
+        dropout_rate (float): dropout rate
+
+    Returns:
+        (function): function to create Transformer layer
+
+    """
+    if not {"d_hidden", "d_ff", "heads"} <= block_arch.keys():
+        logging.warning(
+            "Transformer layer format is: {'type: transformer', "
+            "'d_hidden': int, 'd_ff': int, 'heads': int}"
+        )
+        exit(1)
```

Comment:
Instead of `exit(1)`, how about `raise ValueError("Transformer layer format is ...")`. This can provide where the error happened.